File: 50-powx-n/powx-n.py
class Solution(object):
    def myPow(self, x, n):
        """
        :type x: float
        :type n: int
        :rtype: float
        """
        sign = True
        if n < 0 : sign = False
        n = abs(n)
        ans = 1.0

        while n > 0:
            if n == 1:
                ans = ans * x
                if sign == True:
                    return ans
                return 1/ans

            if n % 2 == 0:
                x = x * x
                n = n / 2
            else:
                ans = ans * x
                n = n - 1
        
        return ans



        

File: 50-powx-n/powx-n.cpp
class Solution {
private:
    double get(double x, int n){
        if(n == 1) return x;
        if(n == 0) return 1;

        double half = get(x, n / 2);
        if(n % 2 == 0) return half * half;
        return half * half * x;
    }
public:
    double myPow(double x, int n) {

        bool sign = true;
        if(n < 0) sign = false;
        double ans = get(x, n);
        if (sign) return ans;
        return 1/ans;
    }
};